Sources/MapboxNavigation/CarPlaySearchController.swift

```diff
@@ -1,13 +1,123 @@
+    }
+    
+    public init (from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        title = try container.decode(String.self, forKey: .title)
+        address = try container.decode(String.self, forKey: .address)
+        if let locationHolder = try container.decodeIfPresent(CLLocationModel.self, forKey: .location) {
+            location = CLLocation(model: locationHolder)
+        }
+        if let routableLocationsHolder = try container.decodeIfPresent([CLLocationModel].self, forKey: .routableLocations) {
+            routableLocations = routableLocationsHolder.map { CLLocation(model: $0) }
+        }
+    }
+    
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        try container.encode(title, forKey: .title)
+        try container.encodeIfPresent(address, forKey: .address)
+        try container.encodeIfPresent(location, forKey: .location)
+        try container.encodeIfPresent(routableLocations, forKey: .routableLocations)
+    }
+    
+    // SHOULD this equality be more strict?
+    public static func == (lhs: NavGeocodedPlacemark, rhs: NavGeocodedPlacemark) -> Bool {
+        return lhs.title == rhs.title &&
+            lhs.address == rhs.address
+    }
+    
+    @available(iOS 12.0, *)
+    public func listItem() -> CPListItem {
+        let item = CPListItem(text: title, detailText: address, image: nil, showsDisclosureIndicator: true)
+        item.userInfo = [CarPlaySearchController.CarPlayGeocodedPlacemarkKey: self]
+        return item
+    }
+}
+
+extension CLLocation: Encodable {
+    public enum CodingKeys: String, CodingKey {
+        case latitude
+        case longitude
+        case altitude
+        case horizontalAccuracy
+        case verticalAccuracy
+        case speed
+        case course
+        case timestamp
+    }
+    
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        try container.encode(coordinate.latitude, forKey: .latitude)
+        try container.encode(coordinate.longitude, forKey: .longitude)
+        try container.encode(altitude, forKey: .altitude)
+        try container.encode(horizontalAccuracy, forKey: .horizontalAccuracy)
+        try container.encode(verticalAccuracy, forKey: .verticalAccuracy)
+        try container.encode(speed, forKey: .speed)
+        try container.encode(course, forKey: .course)
+        try container.encode(timestamp, forKey: .timestamp)
+    }
+    
+    convenience init(model: CLLocationModel) {
+        self.init(coordinate: CLLocationCoordinate2DMake(model.latitude, model.longitude), altitude: model.altitude, horizontalAccuracy: model.horizontalAccuracy, verticalAccuracy: model.verticalAccuracy, course: model.course, speed: model.speed, timestamp: model.timestamp)
+    }
+}
+
+struct CLLocationModel: Codable {
```

Comment:
Is this a workaround for the fact that CLLocation isn’t Codable? If so, it may be simpler to extend CLLocation with Codable conformance.

Reply:
I extended CLLocation to conform to Encodable, but non-convenience initializers must be declared directly in the CLLocation class and not within an extension. Personally, the elegant solution was creating a convenience initializer using a model instead of decoding each CLLocation property individually.